File: src/easy/383-ransom-note/ransomNote.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        map = {}

        for x in ransomNote:
          if x in map:
            map[x] = map[x] + 1
          else:
            map[x] = 1

        for x in magazine:
          if x in map:
            map[x] = map[x] - 1
        
        return all(i <= 0 for i in map.values())

testSolution = Solution()

print(testSolution.canConstruct("a", "b"));                                                        # output == false
print(testSolution.canConstruct("aa", "ab"));                                                      # output == false
print(testSolution.canConstruct("aa", "aab"));                                                     # output == true
print(testSolution.canConstruct("bg", "efjbdfbdgfjhhaiigfhbaejahgfbbgbjagbddfgdiaigdadhcfcj"));    # output == true

File: src/easy/383-ransom-note/ransomNote.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function(ransomNote, magazine) {
  const map = {};

  for (let i = 0; i < ransomNote.length; i++) {
    if (!map[ransomNote[i]]) {
      map[ransomNote[i]] = 1;
    } else {
      map[ransomNote[i]] = map[ransomNote[i]] + 1;
    }
  }

  for (let i = 0; i < magazine.length; i++) {
    if (map[magazine[i]]) {
      map[magazine[i]] = map[magazine[i]] - 1;
    }
  }

  for (const key in map) {
    if (map[key] > 0) {
      return false;
    }
  }

  return true;
};

console.log(canConstruct("a", "b"));                                                        // output == false
console.log(canConstruct("aa", "ab"));                                                      // output == false
console.log(canConstruct("aa", "aab"));                                                     // output == true
console.log(canConstruct("bg", "efjbdfbdgfjhhaiigfhbaejahgfbbgbjagbddfgdiaigdadhcfcj"));    // output == true
